File: validpalindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        s1 = ""
        if (len(s) == 0) or (len(s) == 1):
            return True

        #filter
        for i in range(0, len(s)):
            if s[i].isalnum() == True:
                s1 = s1 + s[i].lower()
            else:
                continue

        left = 0
        right = len(s1) - 1

        for i in range(0, int(len(s1)/2)):
            if s1[left] != s1[right]:
                return False
            else:
                left+=1
                right-=1
        
        return True
        
#2 pointer solution: time complexity = O(n)

File: validpalindrome.java
class Solution {
    public boolean isPalindrome(String s) {
        if(s.equals(""))
            return true;
        String s1 = s.toLowerCase().replaceAll("[^0-9a-z]", "");
        String[] s1Array = s1.split("", 0);
        int left = 0;
        int right = s1Array.length - 1;
        for(int i = 0; i < s1Array.length / 2; i++){
            if(!s1Array[left].equals(s1Array[right])){
                return false;
            }
            ++left;
            --right;
        }
        
        return true;
    }
}

//using 2 pointers => tolowercase = O(n), replaceall = O(n), split = O(n)
//for loop = O(n/2) => total time complexity = O(n)
